compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1838,19 +1893,36 @@ impl Type {
     /// If this type is a Type::Constant (used in array lengths), or is bound
     /// to a Type::Constant, return the constant as a u32.
     pub fn evaluate_to_u32(&self) -> Option<u32> {
-        if let Some((binding, _kind)) = self.get_inner_type_variable() {
+        self.evaluate_to_field_element(&Kind::u32())
+            .and_then(|field_element| field_element.try_to_u32())
+    }
+
+    pub(crate) fn evaluate_to_field_element(&self, kind: &Kind) -> Option<acvm::FieldElement> {
+        if let Some((binding, binding_kind)) = self.get_inner_type_variable() {
             if let TypeBinding::Bound(binding) = &*binding.borrow() {
-                return binding.evaluate_to_u32();
+                if kind.unifies(&binding_kind) {
+                    return binding.evaluate_to_field_element(&binding_kind);
+                }
```

Comment:
Hopefully we can find out where we're missing kind checks elsewhere so we can remove these kind checks during evaluation eventually. That's outside this PR though.

Reply:
Added a follow up issue and TODO for it: https://github.com/noir-lang/noir/issues/6260